access/ms-teams/app.go

```diff
@@ -0,0 +1,517 @@
+			log.WithError(err).Errorf("Failed to process deleted request")
+			return trace.Wrap(err)
+		}
+		return nil
+	default:
+		return trace.BadParameter("unexpected event operation %s", op)
+	}
+}
+
+// onPendingRequest is called when there's a new request/update to pending request
+func (a *App) onPendingRequest(ctx context.Context, req types.AccessRequest) error {
+	log := logger.Get(ctx)
+
+	id := req.GetName()
+	data := pd.AccessRequestData{
+		User:          req.GetUser(),
+		Roles:         req.GetRoles(),
+		RequestReason: req.GetRequestReason(),
+	}
+
+	pd, err := a.pd.Modify(ctx, id, func(existing interface{}) (interface{}, bool) {
+		if existing != nil {
+			return PluginData{}, false
+		}
+		return PluginData{AccessRequestData: data}, true
+	})
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	// If we had not sent any messages before, let's create them
+	if pd != nil {
+		recipients := a.getMessageRecipients(ctx, req)
+
+		if len(recipients) == 0 {
+			log.Warning("No recipients to notify")
+		} else {
+			err = a.postMessages(ctx, recipients, id, data)
+			if err != nil {
+				return trace.Wrap(err)
+			}
+		}
+	}
+
+	// Means that we are updating the request
+	reviews := req.GetReviews()
+	if len(reviews) == 0 {
+		return nil
+	}
+
+	err = a.postReviews(ctx, id, reviews)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	return nil
+}
+
+// onResolvedRequest is called when a request is resolved
+func (a *App) onResolvedRequest(ctx context.Context, req types.AccessRequest) error {
+	replyErr := a.postReviews(ctx, req.GetName(), req.GetReviews())
+
+	var tag pd.ResolutionTag
+	state := req.GetState()
+	switch state {
+	case types.RequestState_APPROVED:
+		tag = pd.ResolvedApproved
+	case types.RequestState_DENIED:
+		tag = pd.ResolvedDenied
+	default:
+		logger.Get(ctx).Warningf("Unknown state %v (%s)", state, state.String())
+		return trace.Wrap(replyErr)
+	}
+	err := a.updateMessages(ctx, req.GetName(), tag, req.GetResolveReason(), req.GetReviews())
+	if replyErr != nil || err != nil {
+		return trace.NewAggregate(replyErr, err)
+	}
+	return nil
+}
+
+// onDeleteRequest gets called when a request is deleted
+func (a *App) onDeletedRequest(ctx context.Context, reqID string) error {
+	return a.updateMessages(ctx, reqID, pd.ResolvedExpired, "", nil)
+}
+
+// postMessages posts initial Teams messages
+func (a *App) postMessages(ctx context.Context, recipients []string, id string, data pd.AccessRequestData) error {
+	teamsData, err := a.bot.PostMessages(ctx, recipients, id, data)
+	if len(teamsData) == 0 && err != nil {
+		return trace.Wrap(err)
+	}
+	for _, data := range teamsData {
+		logger.Get(ctx).WithFields(logger.Fields{
+			"id":        data.ID,
+			"timestamp": data.Timestamp,
+			"recipient": data.RecipientID,
+		}).Info("Successfully posted to MS Teams")
+	}
+	if err != nil {
+		logger.Get(ctx).WithError(err).Error("Failed to post one or more messages to MS Teams")
+	}
```

Comment:
nit: check the error in one go right after the call
```suggestion
	teamsData, err := a.bot.PostMessages(ctx, recipients, id, data)
	if err != nil {
		len(teamsData) == 0 {
			return trace.Wrap(err)
		}
		logger.Get(ctx).WithError(err).Error("Failed to post one or more messages to MS Teams")
	}
	for _, data := range teamsData {
		logger.Get(ctx).WithFields(logger.Fields{
			"id":        data.ID,
			"timestamp": data.Timestamp,
			"recipient": data.RecipientID,
		}).Info("Successfully posted to MS Teams")
	}
```